### PR TITLE
Fix external module encoding

### LINKED ISSUE #15577: Auxiliary failed: Encoding::UndefinedConversionError "\xE4" from ASCII-8BIT to UTF-8
## Steps to reproduce

How'd you do it?

1. writes a python module
```python
#!/usr/bin/env python3


import json

from metasploit import module

metadata = {
    'name': 'test',
    'description': '1',
    'authors': ['1'],
    'date': '2021-08-20',
    'references': [],
    'type': 'dos',
    'options': {
        'query': {'type': 'string', 'description': 'search something', 'required': False, 'default': ''},
    }}


def run(args):
    module.log(args["query"])


if __name__ == "__main__":
    module.run(metadata, run)
```
2. puts it in `~/.msf4/modules/auxiliary/gather`, if the dir does not exists, creates it.
3.  run msfconsole and set query with Chinese 
4. environment     
os: macos 11.5    
version:
Framework: 6.1.1-dev-d53cc717ba971b0e01fa46aadf93bf013f7f3f4c
Console  : 6.1.1-dev-d53cc717ba971b0e01fa46aadf93bf013f7f3f4c
5. more logs
```
[-] Auxiliary failed: Encoding::UndefinedConversionError "\xE4" from ASCII-8BIT to UTF-8
[-] Call stack:
[-]   /opt/metasploit-framework/embedded/lib/ruby/gems/2.7.0/gems/json-2.5.1/lib/json/common.rb:312:in `generate'
[-]   /opt/metasploit-framework/embedded/lib/ruby/gems/2.7.0/gems/json-2.5.1/lib/json/common.rb:312:in `generate'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/message.rb:37:in `to_json'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/bridge.rb:70:in `send'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/bridge.rb:18:in `exec'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external.rb:25:in `exec'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/module/external.rb:8:in `execute_module'
```

## Expected behavior
work normally

## Current behavior
image

### Metasploit version
Framework: 6.1.1-dev-d53cc717ba971b0e01fa46aadf93bf013f7f3f4c
Console  : 6.1.1-dev-d53cc717ba971b0e01fa46aadf93bf013f7f3f4c

### PR DESCRIPTION
**Pull request by sjanusz-r7**
Fixes #15577

This PR properly handles encoding of UTF-8 characters.

It also makes a small correction to one of the relevant Python modules which was necessary to run the module.

### Before

```
[-] Auxiliary failed: Encoding::UndefinedConversionError "\xE4" from ASCII-8BIT to UTF-8
[-] Call stack:
[-]   /opt/metasploit-framework/embedded/lib/ruby/gems/2.7.0/gems/json-2.5.1/lib/json/common.rb:312:in `generate'
[-]   /opt/metasploit-framework/embedded/lib/ruby/gems/2.7.0/gems/json-2.5.1/lib/json/common.rb:312:in `generate'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/message.rb:37:in `to_json'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/bridge.rb:70:in `send'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external/bridge.rb:18:in `exec'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/modules/external.rb:25:in `exec'
[-]   /opt/metasploit-framework/embedded/framework/lib/msf/core/module/external.rb:8:in `execute_module'
```

### After

```
msf6 auxiliary(gather/python_gather) > run

[*] Starting server...
[*] 漢字
[*] Auxiliary module execution completed
```

In this example the UTF-8 encoded characters are simply output to the terminal for testing purposes.

## Verification

- [ ] Follow the steps from #15577

### CODE REVIEW

File: lib/msf/core/data_store.rb [hunk 1]
```diff
@@ -169,29 +169,29 @@ def to_s(delim = ' ')
   def to_h
     datastore_hash = {}
     self.keys.each do |k|
-      datastore_hash[k.to_s] = self[k].to_s
+      datastore_hash[k.to_s.dup.force_encoding('UTF-8')] = self[k].to_s.dup.force_encoding('UTF-8')
```
Comment (adfoster-r7) [inline lib/msf/core/data_store.rb:172]: I believe we can revert this line as only the other method is called as part of external modules? 👀

File: lib/msf/core/data_store.rb [hunk 2]
```diff
@@ -169,29 +169,29 @@ def to_s(delim = ' ')
...
-  def to_nested_values
+  def to_external_message_h
     datastore_hash = {}
 
     array_nester = ->(arr) do
       if arr.first.is_a? Array
         arr.map &array_nester
       else
-        arr.map &:to_s
+        arr.map { |item| item.to_s.dup.force_encoding('UTF-8') } # &:to_s
```
Comment (adfoster-r7) [inline lib/msf/core/data_store.rb:185]: Looks like debugging code is left in here 👀 

```suggestion
        arr.map { |item| item.to_s.dup.force_encoding('UTF-8') }
```

Comment (adfoster-r7) [reply lib/msf/core/data_store.rb:185]: Just confirming if you were wanting to convert `nil` into an empty string here

File: lib/msf/core/data_store.rb [hunk 3]
```diff
@@ -169,29 +169,29 @@ def to_s(delim = ' ')
...
     datastore_hash = {}
     self.keys.each do |k|
-      datastore_hash[k.to_s] = self[k].to_s
+      datastore_hash[k.to_s.dup.force_encoding('UTF-8')] = self[k].to_s.dup.force_encoding('UTF-8')
     end
     datastore_hash
   end
 
   # Hack on a hack for the external modules
-  def to_nested_values
```
Comment (adfoster-r7) [inline lib/msf/core/data_store.rb:178]: We traced down the original use of this method to https://github.com/rapid7/metasploit-framework/pull/10002 which allows datastore values to be a nested array, which I'm not sure is something we'd want

File: lib/msf/core/data_store.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/msf/core/data_store.rb]: As discussed; Let's leave the nil -> empty string conversion here as it might introduce regressions :+1:

**Review by gwillcox-r7 — APPROVED** [review]

**Pull request merged**